internal/webhooks/cluster.go

```diff
@@ -529,24 +532,25 @@ func ValidateClusterForClusterClass(cluster *clusterv1.Cluster, clusterClass *cl
 // pollClusterClassForCluster will retry getting the ClusterClass referenced in the Cluster for two seconds.
 func (webhook *Cluster) pollClusterClassForCluster(ctx context.Context, cluster *clusterv1.Cluster) (*clusterv1.ClusterClass, error) {
 	clusterClass := &clusterv1.ClusterClass{}
-	var clusterClassGetErr error
+	var clusterClassPollErr error
+	// TODO: Add a webhook warning if the ClusterClass is not up to date or not found.
 	_ = util.PollImmediate(200*time.Millisecond, 2*time.Second, func() (bool, error) {
-		if clusterClassGetErr = webhook.Client.Get(ctx, client.ObjectKey{Namespace: cluster.Namespace, Name: cluster.Spec.Topology.Class}, clusterClass); clusterClassGetErr != nil {
+		if clusterClassPollErr = webhook.Client.Get(ctx, client.ObjectKey{Namespace: cluster.Namespace, Name: cluster.Spec.Topology.Class}, clusterClass); clusterClassPollErr != nil {
 			return false, nil //nolint:nilerr
 		}
 
-		// Return an error if the ClusterClass has not successfully reconciled because variables aren't correctly
-		// reconciled.
-		// TODO: Decide whether to check generation here. This requires creating templates before creating the Cluster and
-		// may interfere with the way clusterctl move works.
-		if !conditions.Has(clusterClass, clusterv1.ClusterClassVariablesReconciledCondition) ||
+		// Return errClusterClassNotReconciled if the ClusterClass has not successfully reconciled or if the
+		// ClusterClass variables have not been successfully reconciled.:
+		if clusterClass.Generation != clusterClass.Status.ObservedGeneration ||
+			!conditions.Has(clusterClass, clusterv1.ClusterClassVariablesReconciledCondition) ||
 			conditions.IsFalse(clusterClass, clusterv1.ClusterClassVariablesReconciledCondition) {
-			clusterClassGetErr = errors.New("ClusterClass is not up to date. If this persists check ClusterClass status")
+			clusterClassPollErr = errClusterClassNotReconciled
```

Comment:
Found out what the error an apply was all about:

The following happens:
* ClusterClass controller is reconciling, in defer
  1. patch helper is writing status conditions
  2. patch helper is writing spec
  3. patch helper is writing status (including variables)
* e2e test code applies YAML (with ClusterClass and Cluster) in a way that the Cluster is applied between 1 and 3
  * Result: conditions and observedGeneration are all good, but the variables do not exist yet => validation error from the webhook
  
Bonus: e2e test code is not retrying because we only set a timeout but not a retry interval: https://github.com/kubernetes-sigs/cluster-api/blob/df94b2a19d97a2e3bfc278cfc72582b08e4459d7/test/framework/clusterctl/clusterctl_helpers.go#L301 (no idea if that is intentional, based on commit message it was not) 

I think if we don't want to refactor the patch helper (now) we also have to check if the ClusterClass.status.variables exist (on top of checking generation & VariablesDiscovered condition)

(@killianmuldoon)

Reply:
+1 to add the additional check.
Let's just consider to move all the conditions in this if into a func called `IsClusterClassRecondiled` or `IsClusterClassReady`

Reply:
Sounds good!